## APARTMENT RENTAL IN GERMNAY: Consistency & Wrangling

### 6.1 Sourcing open data

### Importing Libraries & Dataframe

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
path=r'C:\Users\Jatin\Documents\ApartmentRentalInGermany\Data'

In [3]:
# Importing dataframe
df=pd.read_csv(os.path.join(path,'Original Data','immo_data.csv'))

In [4]:
pd.options.display.max_columns=None

### Conducting basic checks

In [5]:
# Checking dimensions
df.shape

(268850, 49)

In [6]:
# Checking basic data
df.head()

regio1  serviceCharge                     heatingType  \
0  Nordrhein_Westfalen         245.00                 central_heating   
1      Rheinland_Pfalz         134.00  self_contained_central_heating   
2              Sachsen         255.00                   floor_heating   
3              Sachsen          58.15                district_heating   
4               Bremen         138.00  self_contained_central_heating   

  telekomTvOffer  telekomHybridUploadSpeed  newlyConst  balcony  picturecount  \
0  ONE_YEAR_FREE                       NaN       False    False             6   
1  ONE_YEAR_FREE                       NaN       False     True             8   
2  ONE_YEAR_FREE                      10.0        True     True             8   
3  ONE_YEAR_FREE                       NaN       False     True             9   
4            NaN                       NaN       False     True            19   

   pricetrend  telekomUploadSpeed  totalRent  yearConstructed    scoutId  \
0        4.62                10.0      840.0           1965.0   96107057   
1        3.47                10.0        NaN           1871.0  111378734   
2        2.72                 2.4     1300.0           2019.0  113147523   
3        1.53                40.0        NaN           1964.0  108890903   
4        2.46                 NaN      903.0           1950.0  114751222   

   noParkSpaces       firingTypes  hasKitchen              geo_bln  cellar  \
0           1.0               oil       False  Nordrhein_Westfalen    True   
1           2.0               gas       False      Rheinland_Pfalz   False   
2           1.0               NaN       False              Sachsen    True   
3           NaN  district_heating       False              Sachsen   False   
4           NaN               gas       False               Bremen   False   

   yearConstructedRange  baseRent houseNumber  livingSpace  \
0                   2.0     595.0         244        86.00   
1                   1.0     800.0         NaN        89.00   
2                   9.0     965.0           4        83.80   
3                   2.0     343.0          35        58.15   
4                   1.0     765.0          10        84.97   

               geo_krs       condition   interiorQual petsAllowed  \
0             Dortmund       well_kept         normal         NaN   
1    Rhein_Pfalz_Kreis     refurbished         normal          no   
2              Dresden  first_time_use  sophisticated         NaN   
3  Mittelsachsen_Kreis             NaN            NaN         NaN   
4               Bremen     refurbished            NaN         NaN   

                        street                  streetPlain   lift  \
0   Sch&uuml;ruferstra&szlig;e              Schüruferstraße  False   
1               no_information                          NaN  False   
2                    Turnerweg                    Turnerweg   True   
3  Gl&uuml;ck-Auf-Stra&szlig;e             Glück-Auf-Straße  False   
4  Hermann-Henrich-Meier-Allee  Hermann-Henrich-Meier-Allee  False   

   baseRentRange    typeOfFlat  geo_plz  noRooms  thermalChar  floor  \
0              4  ground_floor    44269      4.0        181.4    1.0   
1              5  ground_floor    67459      3.0          NaN    NaN   
2              6     apartment     1097      3.0          NaN    3.0   
3              2         other     9599      3.0         86.0    3.0   
4              5     apartment    28213      3.0        188.9    1.0   

   numberOfFloors  noRoomsRange  garden  livingSpaceRange  \
0             3.0             4    True                 4   
1             NaN             3   False                 4   
2             4.0             3   False                 4   
3             NaN             3   False                 2   
4             NaN             3   False                 4   

                regio2                      regio3  \
0             Dortmund                     Schüren   
1    Rhein_Pfalz_Kreis              Böhl_Iggelh

### DATA WRANGLING

### a. Dropping columns

In [7]:
# Checking for missing values
df.isnull().sum()

regio1                           0
serviceCharge                 6909
heatingType                  44856
telekomTvOffer               32619
telekomHybridUploadSpeed    223830
newlyConst                       0
balcony                          0
picturecount                     0
pricetrend                    1832
telekomUploadSpeed           33358
totalRent                    40517
yearConstructed              57045
scoutId                          0
noParkSpaces                175798
firingTypes                  56964
hasKitchen                       0
geo_bln                          0
cellar                           0
yearConstructedRange         57045
baseRent                         0
houseNumber                  71018
livingSpace                      0
geo_krs                          0
condition                    68489
interiorQual                112665
petsAllowed                 114573
street                           0
streetPlain                  71013
lift                

In [8]:
# Dropping columns which are not necessary for analysis
df_drop=df.drop(columns=['telekomTvOffer','telekomHybridUploadSpeed','picturecount','pricetrend','telekomUploadSpeed','scoutId','noParkSpaces','firingTypes','geo_bln','yearConstructedRange','houseNumber','interiorQual','petsAllowed','street','streetPlain','baseRentRange','thermalChar','numberOfFloors','noRoomsRange','livingSpaceRange','geo_krs','description','facilities','heatingCosts','energyEfficiencyClass','lastRefurbish','electricityBasePrice','electricityKwhPrice','geo_plz'])

In [9]:
df_drop.shape

(268850, 20)

In [10]:
df_drop.head()

regio1  serviceCharge                     heatingType  \
0  Nordrhein_Westfalen         245.00                 central_heating   
1      Rheinland_Pfalz         134.00  self_contained_central_heating   
2              Sachsen         255.00                   floor_heating   
3              Sachsen          58.15                district_heating   
4               Bremen         138.00  self_contained_central_heating   

   newlyConst  balcony  totalRent  yearConstructed  hasKitchen  cellar  \
0       False    False      840.0           1965.0       False    True   
1       False     True        NaN           1871.0       False   False   
2        True     True     1300.0           2019.0       False    True   
3       False     True        NaN           1964.0       False   False   
4       False     True      903.0           1950.0       False   False   

   baseRent  livingSpace       condition   lift    typeOfFlat  noRooms  floor  \
0     595.0        86.00       well_kept  False  ground_floor      4.0    1.0   
1     800.0        89.00     refurbished  False  ground_floor      3.0    NaN   
2     965.0        83.80  first_time_use   True     apartment      3.0    3.0   
3     343.0        58.15             NaN  False         other      3.0    3.0   
4     765.0        84.97     refurbished  False     apartment      3.0    1.0   

   garden               regio2                      regio3   date  
0    True             Dortmund                     Schüren  May19  
1   False    Rhein_Pfalz_Kreis              Böhl_Iggelheim  May19  
2   False              Dresden  Äußere_Neustadt_Antonstadt  Oct19  
3   False  Mittelsachsen_Kreis                    Freiberg  May19  
4   False               Bremen           Neu_Schwachhausen  Feb20

### b. Renaming columns

In [11]:
# renaming some columns for better understanding
df_drop.rename(columns={'regio1':'state','regio2':'district','regio3':'city','noRooms':'numberOfRooms','newlyConst':'newlyConstructed'},inplace=True)

In [12]:
df_drop.head(60)

state  serviceCharge                     heatingType  \
0      Nordrhein_Westfalen         245.00                 central_heating   
1          Rheinland_Pfalz         134.00  self_contained_central_heating   
2                  Sachsen         255.00                   floor_heating   
3                  Sachsen          58.15                district_heating   
4                   Bremen         138.00  self_contained_central_heating   
5       Schleswig_Holstein         142.00                     gas_heating   
6                  Sachsen          70.00  self_contained_central_heating   
7                   Bremen          88.00                 central_heating   
8        Baden_Württemberg         110.00                     oil_heating   
9      Nordrhein_Westfalen          95.00  self_contained_central_heating   
10                 Sachsen          88.00                             NaN   
11                 Sachsen         155.00                 central_heating   
12         Rheinland_Pfalz         270.00                     oil_heating   
13     Nordrhein_Westfalen         200.00                 central_heating   
14     Nordrhein_Westfalen         215.00                     gas_heating   
15               Thüringen          80.00                district_heating   
16                  Hessen         220.00                     oil_heating   
17                 Sachsen          50.00                 central_heating   
18     Nordrhein_Westfalen         121.00                 central_heating   
19           Niedersachsen         290.00                district_heating   
20      Schleswig_Holstein         140.00                 central_heating   
21       Baden_Württemberg           0.00                             NaN   
22                 Sachsen         346.00                 central_heating   
23                  Bayern         180.00                 central_heating   
24                 Hamburg         140.00                 central_heating   
25           Niedersachsen          80.00                district_heating   
26       Baden_Württemberg          20.00                             NaN   
27          Sachsen_Anhalt         147.32                district_heating   
28                  Bayern           0.00                             NaN   
29  Mecklenburg_Vorpommern         104.02                district_heating   
30                  Bayern         170.00                   floor_heating   
31     Nordrhein_Westfalen         130.00                district_heating   
32                  Bayern           0.00                             NaN   
33                 Sachsen          80.00                 central_heating   
34                  Hessen         260.00             wood_pellet_heating   
35                 Sachsen         140.00                 central_heating   
36                 Sachsen         140.00                district_heating   
37     Nordrhein_Westfalen         100.00                 central_heating   
38     Nordrhein_Westfalen         150.00                             NaN   
39                 Sachsen         103.70                 central_heating   
40       Baden_Württemberg         230.00                 central_heating   
41           Niedersachsen         196.80                             NaN   
42     Nordrhein_Westfalen         200.00                district_heating   
43     Nordrhein_Westfalen         160.00                 central_heating   
44                 Sachsen         130.00                     oil_heating   
45                  Berlin         320.00                 central_heating   
46                  Hessen           0.00                             NaN   
47          Sachsen_Anhalt          80.00                 central_heating   
48      Schleswig_Holstein         121.50                             NaN   
49                  Bayern         185.00                             NaN   
50             Brandenburg          50.00                 central_heating   
51               

### c. Changing data types

In [13]:
# Changing data type from float to int
df_drop['numberOfRooms']=df_drop['numberOfRooms'].astype('int64')

In [14]:
df_drop['numberOfRooms'].dtype

dtype('int64')

## DATA CONSISTENCY CHECKS

### a. Checking for mixed type data

In [15]:
# Checking for mixed type data
for col in df_drop.columns.tolist():
  weird = (df_drop[[col]].applymap(type) != df_drop[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_drop[weird]) > 0:
    print (col)

heatingType
condition
typeOfFlat


In [16]:
# Updating heatingType column data type to string
df_drop['heatingType']=df_drop['heatingType'].astype('str')

In [17]:
# Updating condition column data type to string
df_drop['condition']=df_drop['condition'].astype('str')

In [18]:
# Updating typeOfFlat column data type to string
df_drop['typeOfFlat']=df_drop['typeOfFlat'].astype('str')

In [19]:
# Checking again for mixed type data
for col in df_drop.columns.tolist():
  weird = (df_drop[[col]].applymap(type) != df_drop[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_drop[weird]) > 0:
    print (col)

NO mixed type data found.

## b. Checking for missing values

In [20]:
# Checking for missing values
df_drop.isnull().sum()

state                   0
serviceCharge        6909
heatingType             0
newlyConstructed        0
balcony                 0
totalRent           40517
yearConstructed     57045
hasKitchen              0
cellar                  0
baseRent                0
livingSpace             0
condition               0
lift                    0
typeOfFlat              0
numberOfRooms           0
floor               51309
garden                  0
district                0
city                    0
date                    0
dtype: int64

In [21]:
# Checking descriptive Statistics
df_drop.describe()

serviceCharge     totalRent  yearConstructed      baseRent  \
count  261941.000000  2.283330e+05    211805.000000  2.688500e+05   
mean      151.206113  9.013315e+02      1966.400590  6.941294e+02   
std       308.295790  3.323833e+04        46.992207  1.953602e+04   
min         0.000000  0.000000e+00      1000.000000  0.000000e+00   
25%        95.000000  4.698000e+02      1950.000000  3.380000e+02   
50%       135.000000  6.500000e+02      1973.000000  4.900000e+02   
75%       190.000000  9.850000e+02      1996.000000  7.990000e+02   
max    146118.000000  1.575154e+07      2090.000000  9.999999e+06   

         livingSpace  numberOfRooms          floor  
count  268850.000000  268850.000000  217541.000000  
mean       74.355548       2.595808       2.122405  
std       254.759208       2.630466       3.634934  
min         0.000000       1.000000      -1.000000  
25%        54.000000       2.000000       1.000000  
50%        67.320000       3.000000       2.000000  
75%        87.000000       3.000000       3.000000  
max    111111.000000     999.000000     999.000000

In [22]:
# To calculate median 
df_drop.median()

C:\Users\Jatin\AppData\Local\Temp\ipykernel_19760\3299781122.py:2: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_drop.median()


serviceCharge        135.00
newlyConstructed       0.00
balcony                1.00
totalRent            650.00
yearConstructed     1973.00
hasKitchen             0.00
cellar                 1.00
baseRent             490.00
livingSpace           67.32
lift                   0.00
numberOfRooms          3.00
floor                  2.00
garden                 0.00
dtype: float64

In [23]:
## replacing missing values with its mean value
df_drop['serviceCharge'].fillna(151,inplace=True)

In [24]:
# # replacing missing values with its mean value
df_drop['totalRent'].fillna(902,inplace=True)

In [25]:
# replacing missing values with its median value
df_drop['yearConstructed'].fillna(1973,inplace=True)

In [26]:
# replacing missing values with its median value
df_drop['floor'].fillna(2,inplace=True)

In [27]:
# Checking for missing values
df_drop.isnull().sum()

state               0
serviceCharge       0
heatingType         0
newlyConstructed    0
balcony             0
totalRent           0
yearConstructed     0
hasKitchen          0
cellar              0
baseRent            0
livingSpace         0
condition           0
lift                0
typeOfFlat          0
numberOfRooms       0
floor               0
garden              0
district            0
city                0
date                0
dtype: int64

 All missing values are replaced by mean or median of selected columns.

In [28]:
df_drop.head(10)

state  serviceCharge                     heatingType  \
0  Nordrhein_Westfalen         245.00                 central_heating   
1      Rheinland_Pfalz         134.00  self_contained_central_heating   
2              Sachsen         255.00                   floor_heating   
3              Sachsen          58.15                district_heating   
4               Bremen         138.00  self_contained_central_heating   
5   Schleswig_Holstein         142.00                     gas_heating   
6              Sachsen          70.00  self_contained_central_heating   
7               Bremen          88.00                 central_heating   
8    Baden_Württemberg         110.00                     oil_heating   
9  Nordrhein_Westfalen          95.00  self_contained_central_heating   

   newlyConstructed  balcony  totalRent  yearConstructed  hasKitchen  cellar  \
0             False    False     840.00           1965.0       False    True   
1             False     True     902.00           1871.0       False   False   
2              True     True    1300.00           2019.0       False    True   
3             False     True     902.00           1964.0       False   False   
4             False     True     903.00           1950.0       False   False   
5             False     True     902.00           1999.0        True   False   
6             False    False     380.00           1973.0       False    True   
7             False     True     584.25           1959.0       False    True   
8             False    False     690.00           1970.0        True    True   
9             False    False     902.00           1953.0       False    True   

   baseRent  livingSpace        condition   lift    typeOfFlat  numberOfRooms  \
0    595.00        86.00        well_kept  False  ground_floor              4   
1    800.00        89.00      refurbished  False  ground_floor              3   
2    965.00        83.80   first_time_use   True     apartment              3   
3    343.00        58.15              nan  False         other              3   
4    765.00        84.97      refurbished  False     apartment              3   
5    315.20        53.43        well_kept  False     apartment              2   
6    310.00        62.00  fully_renovated  False           nan              2   
7    452.25        60.30              nan  False  ground_floor              3   
8    580.00        53.00        well_kept  False   roof_storey              2   
9    300.00        60.00        well_kept  False     apartment              2   

   floor  garden                   district                        city   date  
0    1.0    True                   Dortmund                     Schüren  May19  
1    2.0   False          Rhein_Pfalz_Kreis              Böhl_Iggelheim  May19  
2    3.0   False                    Dresden  Äußere_Neustadt_Antonstadt  Oct19  
3    3.0   False        Mittelsachsen_Kreis                    Freiberg  May19  
4    1.0   False                     Bremen           Neu_Schwachhausen  Feb20  
5    2.0   False  Schleswig_Flensburg_Kreis                   Struxdorf  Feb20  
6    1.0    True        Mittelsachsen_Kreis                    Freiberg  Feb20  
7    2.0   False                     Bremen                  St._Magnus  Feb20  
8    2.0   False          Emmendingen_Kreis                  Denzlingen  Feb20  
9    2.0   False              Gelsenkirchen               Bulmke_Hüllen  Oct19

In [29]:
df_drop.shape

(268850, 20)

### c. Checking for duplicate values

In [30]:
# Creating a subset that contain only rows with duplicate values
df_dups = df_drop[df_drop.duplicated()]

In [31]:
df_dups

state  serviceCharge       heatingType  \
2811                   Sachsen         220.00   central_heating   
3114                   Sachsen         121.00   central_heating   
7555                   Sachsen          76.00   central_heating   
9410                   Sachsen         260.00  district_heating   
11601      Nordrhein_Westfalen          35.00               nan   
...                        ...            ...               ...   
268464       Baden_Württemberg         350.00         heat_pump   
268560      Schleswig_Holstein          80.40  district_heating   
268577       Baden_Württemberg         300.00  district_heating   
268786                 Sachsen         255.43               nan   
268801  Mecklenburg_Vorpommern          68.00       gas_heating   

        newlyConstructed  balcony  totalRent  yearConstructed  hasKitchen  \
2811               False    False     678.00           1920.0        True   
3114               False    False     371.00           1973.0        True   
7555               False    False     485.00           1989.0       False   
9410                True     True    1295.00           2019.0       False   
11601              False    False     902.00           1973.0       False   
...                  ...      ...        ...              ...         ...   
268464              True     True    1800.00           2018.0        True   
268560             False    False     389.23           1998.0       False   
268577             False    False    2660.00           1871.0        True   
268786              True     True    1123.88           2019.0       False   
268801             False     True     258.00           1999.0       False   

        cellar  baseRent  livingSpace        condition   lift    typeOfFlat  \
2811     False    458.00        62.35      refurbished  False           nan   
3114      True    250.00        54.91  fully_renovated  False           nan   
7555      True    332.00        63.70   mint_condition  False     apartment   
9410     False   1035.00       129.26   first_time_use   True    maisonette   
11601    False    620.00        44.00              nan  False     apartment   
...        ...       ...          ...              ...    ...           ...   
268464    True   1450.00       135.00   first_time_use  False     apartment   
268560    True    246.83        40.99              nan  False  ground_floor   
268577    True   2200.00        87.00  fully_renovated   True     apartment   
268786   False    868.45       102.17   mint_condition  False     apartment   
268801    True    190.00        28.54       negotiable  False     apartment   

        numberOfRooms  floor  garden                    district  \
2811                2    2.0   False                     Leipzig   
3114                2    1.0    True                    Chemnitz   
7555                3    4.0   False               Bautzen_Kreis   
9410                4    1.0    True                     Leipzig   
11601               2    1.0   False                 Kleve_Kreis   
...               ...    ...     ...                         ...   
268464              5    1.0   False              Tübingen_Kreis   
268560              1    0.0   False                  Neumünster   
268577              2    1.0   False                  Heidelberg   
268786              4    2.0   False                     Leipzig   
268801              1    1.0   False  Mecklenburg_Strelitz_Kreis   

                     city   date  
2811           Mockau_Süd  May19  
3114          Hilbersdorf  Oct19  
7555               Kamenz  May19  
9410    Böhlitz_Ehrenberg  Feb20  
11601               Kleve  Oct19  
...                   ...    ...  
268464       Kusterdingen  May19  
268560         Innenstadt  May19  
268577           Bergheim  Feb20  
268786     Zentrum_Südost  May19  
268801          Friedland  Feb20  

[2939 rows x 20 columns]

2939 duplicate values are found.

In [32]:
# To remove duplicates
df_clean=df_drop.drop_duplicates()

In [33]:
df_clean.shape

(265911, 20)

There were 2939 duplicate values found. It has been dropped and now remaining with 265911 rows.

In [34]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 265911 entries, 0 to 268849
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   state             265911 non-null  object 
 1   serviceCharge     265911 non-null  float64
 2   heatingType       265911 non-null  object 
 3   newlyConstructed  265911 non-null  bool   
 4   balcony           265911 non-null  bool   
 5   totalRent         265911 non-null  float64
 6   yearConstructed   265911 non-null  float64
 7   hasKitchen        265911 non-null  bool   
 8   cellar            265911 non-null  bool   
 9   baseRent          265911 non-null  float64
 10  livingSpace       265911 non-null  float64
 11  condition         265911 non-null  object 
 12  lift              265911 non-null  bool   
 13  typeOfFlat        265911 non-null  object 
 14  numberOfRooms     265911 non-null  int64  
 15  floor             265911 non-null  float64
 16  garden            26

In [35]:
# Changing data type from float to int
df_clean['yearConstructed']=df_clean['yearConstructed'].astype('int64')

C:\Users\Jatin\AppData\Local\Temp\ipykernel_19760\4292057266.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['yearConstructed']=df_clean['yearConstructed'].astype('int64')


In [36]:
df_clean['yearConstructed'].dtype

dtype('int64')

In [37]:
# # Changing data type from float to int
df_clean['floor']=df_clean['floor'].astype('int64')

C:\Users\Jatin\AppData\Local\Temp\ipykernel_19760\3571353479.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['floor']=df_clean['floor'].astype('int64')


In [38]:
df_clean['floor'].dtype

dtype('int64')

In [39]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 265911 entries, 0 to 268849
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   state             265911 non-null  object 
 1   serviceCharge     265911 non-null  float64
 2   heatingType       265911 non-null  object 
 3   newlyConstructed  265911 non-null  bool   
 4   balcony           265911 non-null  bool   
 5   totalRent         265911 non-null  float64
 6   yearConstructed   265911 non-null  int64  
 7   hasKitchen        265911 non-null  bool   
 8   cellar            265911 non-null  bool   
 9   baseRent          265911 non-null  float64
 10  livingSpace       265911 non-null  float64
 11  condition         265911 non-null  object 
 12  lift              265911 non-null  bool   
 13  typeOfFlat        265911 non-null  object 
 14  numberOfRooms     265911 non-null  int64  
 15  floor             265911 non-null  int64  
 16  garden            26

In [40]:
df_clean.head()

state  serviceCharge                     heatingType  \
0  Nordrhein_Westfalen         245.00                 central_heating   
1      Rheinland_Pfalz         134.00  self_contained_central_heating   
2              Sachsen         255.00                   floor_heating   
3              Sachsen          58.15                district_heating   
4               Bremen         138.00  self_contained_central_heating   

   newlyConstructed  balcony  totalRent  yearConstructed  hasKitchen  cellar  \
0             False    False      840.0             1965       False    True   
1             False     True      902.0             1871       False   False   
2              True     True     1300.0             2019       False    True   
3             False     True      902.0             1964       False   False   
4             False     True      903.0             1950       False   False   

   baseRent  livingSpace       condition   lift    typeOfFlat  numberOfRooms  \
0     595.0        86.00       well_kept  False  ground_floor              4   
1     800.0        89.00     refurbished  False  ground_floor              3   
2     965.0        83.80  first_time_use   True     apartment              3   
3     343.0        58.15             nan  False         other              3   
4     765.0        84.97     refurbished  False     apartment              3   

   floor  garden             district                        city   date  
0      1    True             Dortmund                     Schüren  May19  
1      2   False    Rhein_Pfalz_Kreis              Böhl_Iggelheim  May19  
2      3   False              Dresden  Äußere_Neustadt_Antonstadt  Oct19  
3      3   False  Mittelsachsen_Kreis                    Freiberg  May19  
4      1   False               Bremen           Neu_Schwachhausen  Feb20

### EXPORTING DATAFRAME

In [41]:
df.to_csv(os.path.join(path,'Prepared Data','df_clean.csv'))